src/vivarium/framework/values/pipeline.py

```diff
@@ -19,23 +19,58 @@
 class ValueSource(Resource):
     """A resource representing the source of a value pipeline."""
 
-    def __init__(self, name: str) -> None:
-        super().__init__("value_source", name)
-
+    def __init__(self, pipeline: Pipeline, source: Callable[..., Any] | None = None) -> None:
+        super().__init__("value_source" if source else "missing_value_source", pipeline.name)
+        self._pipeline = pipeline
+        self._source = source
 
-class MissingValueSource(Resource):
-    """A resource representing an undefined source of a value pipeline."""
+    def __bool__(self) -> bool:
+        return self._source is not None
 
-    def __init__(self, name: str) -> None:
-        super().__init__("missing_value_source", name)
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        if not self._source:
+            raise DynamicValueError(
+                f"The dynamic value pipeline for {self.name} has no source."
+                " This likely means you are attempting to modify a value that"
+                " hasn't been created."
+            )
+        return self._source(*args, **kwargs)
 
 
 class ValueModifier(Resource):
     """A resource representing a modifier of a value pipeline."""
 
-    def __init__(self, name: str) -> None:
+    def __init__(self, pipeline: Pipeline, modifier: Callable[..., Any]) -> None:
+        mutator_name = self._get_modifier_name(modifier)
+        mutator_index = len(pipeline.mutators) + 1
```

Comment:
Why do we need to add 1 here?

Reply:
Because this is the length of the pipeline's mutators before we've added this `ValueModifier`, and we expect the first one to be named with 1, not 0.